golem_sci/implementation.py

```diff
@@ -251,6 +256,9 @@ def wait_until_synchronized(self) -> bool:
     def is_synchronized(self) -> bool:
         return self._geth_client.is_synchronized()
 
+    def stop(self) -> None:
+        self._monitor_stop.set()
```

Comment:
To really stop everything cleanly we need to wait for monitor to finish.

Reply:
Does this mean that breaking the thread's loop puts the monitor in some unrecoverable state?

Reply:
Breaking the loop is fine. But here you only set the flag, the monitor still might be doing some work. To end cleanly one should wait until the loop is in fact stopped.